lib/learning-object/learning-object.ts

```diff
@@ -71,69 +71,184 @@ export enum AcademicLevel {
    */
-  date: string;
+  get date(): string {
+    return this._date;
+  }
 
+  /**
+   * Updates LearningObject's last-modified date
+   *
+   * @private
+   * @memberof LearningObject
+   */
+  private updateDate() {
+    this._date = Date.now().toString();
+  }
+
+  private _length!: Length;
   /**
    * @property {string} length
    *       the object's class, determining its length (eg. module)
    *       values are restricted according to available lengths
    */
-  length: string;
+  get length(): Length {
+    return this._length;
+  }
+
+  set length(length: Length) {
+    if (this.acceptLength(length)) {
+      this._length = length;
+      this.updateDate();
+    } else {
+      throw new Error(LEARNING_OBJECT_ERRORS.INVALID_LENGTH(length));
+    }
+  }
+
+  /**
+   * Validates length
+   *
+   * @private
+   * @param {Length} length
+   * @returns {boolean}
+   * @memberof LearningObject
+   */
+  private acceptLength(length: Length): boolean {
+    const validLengths: Length[] = [
+      'nanomodule',
+      'micromodule',
+      'module',
+      'unit',
+      'course'
+    ];
+    if (validLengths.includes(length)) {
+      return true;
+    }
+    return false;
+  }
 
+  private _levels: AcademicLevel[];
   /**
    * @property {string[]} levels
-   *       the array of the object's Academic Level. (ie K-12)
+   *       this object's Academic Level. (ie K-12)
    */
-  levels: AcademicLevel[];
+  get levels(): AcademicLevel[] {
+    return this._levels;
+  }
 
   /**
-   * Adds new AcademicLevel to Array of levels if level is not present in the array
+   * Adds new AcademicLevel to array of levels if level is not present in this object's levels
    *
    * @memberof LearningObject
    */
   addLevel(level: AcademicLevel) {
-    if (!this.levels.includes(level)) {
-      this.levels.push(level);
+    const [alreadyAdded, isValid] = this.acceptLevel(level);
+    if (!alreadyAdded && isValid) {
+      this._levels.push(level);
+      this.updateDate();
+    } else if (alreadyAdded) {
+      throw new Error(LEARNING_OBJECT_ERRORS.LEVEL_EXISTS(level));
+    } else {
+      throw new Error(LEARNING_OBJECT_ERRORS.INVALID_LEVEL(level));
     }
   }
+
   /**
-   * Removes AcademicLevel from array of AcademicLevels
+   * Removes AcademicLevel from this object's levels
    *
    * @param {number} index
    * @returns {AcademicLevel}
    * @memberof LearningObject
    */
   removeLevel(index: number): AcademicLevel {
-    return this.levels.splice(index, 1)[0];
+    if (this.levels.length > 1) {
+      this.updateDate();
+      return this._levels.splice(index, 1)[0];
+    } else {
+      throw new Error(LEARNING_OBJECT_ERRORS.INVALID_LEVELS);
+    }
   }
 
+  /**
+   * Validates level and checks if level has already been added
+   *
+   * @private
+   * @param {AcademicLevel} level
+   * @returns {boolean}
+   * @memberof LearningObject
+   */
+  private acceptLevel(level: AcademicLevel): boolean[] {
```

Comment:
I'm not sure this function name is descriptive enough. Perhaps something like isLevelValid or something along those lines. I had to search for the function declaration before I really knew what it did.